ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java

```diff
@@ -219,8 +226,33 @@ public static void mergeDeltaFilesStats(AcidDirectory dir, long checkThresholdIn
     logDeltaDirMetrics(dir, conf, numObsoleteDeltas, numDeltas, numSmallDeltas);
 
     String path = getRelPath(dir);
-    newDeltaFilesStats(numObsoleteDeltas, numDeltas, numSmallDeltas)
-      .forEach((type, cnt) -> deltaFilesStats.computeIfAbsent(type, v -> new HashMap<>()).put(path, cnt));
+
+    filterAndAddToDeltaFilesStats(NUM_DELTAS, numDeltas, deltasThreshold, deltaFilesStats, path, maxCacheSize);
+    filterAndAddToDeltaFilesStats(NUM_OBSOLETE_DELTAS, numObsoleteDeltas, obsoleteDeltasThreshold, deltaFilesStats,
+        path, maxCacheSize);
+    filterAndAddToDeltaFilesStats(NUM_SMALL_DELTAS, numSmallDeltas, deltasThreshold, deltaFilesStats,
+        path, maxCacheSize);
+  }
+
+  /**
+   * Add partition and delta count to deltaFilesStats if the delta count is over the recording threshold and it is in
+   * the top {@link HiveConf.ConfVars#HIVE_TXN_ACID_METRICS_MAX_CACHE_SIZE} deltas.
+   */
+  private static void filterAndAddToDeltaFilesStats(DeltaFilesMetricType type, int deltaCount, int deltasThreshold,
+      EnumMap<DeltaFilesMetricType, Queue<Pair<String, Integer>>> deltaFilesStats, String path, int maxCacheSize) {
+    if (deltaCount > deltasThreshold) {
+      Queue<Pair<String,Integer>> pairQueue = deltaFilesStats.get(type);
+      if (pairQueue != null && pairQueue.size() == maxCacheSize) {
+        Pair<String, Integer> lowest = pairQueue.peek();
+        if (lowest != null && deltaCount > lowest.getValue()) {
+          pairQueue.poll();
+        }
+      }
+      if (pairQueue == null || pairQueue.size() < maxCacheSize) {
+        deltaFilesStats.computeIfAbsent(type,
+            v -> (new PriorityBlockingQueue<>(maxCacheSize, getComparator()))).add(Pair.of(path, deltaCount));
```

Comment:
Why do you need a blocking queue here? Would it be accessed by multiple threads?